// @dagger.io/dagger/dist/src/module/introspector/dagger_module/constructor.js
import { DaggerArgument } from "./argument.js";
export class DaggerConstructor {
    node;
    ast;
    name = "";
    arguments = {};
    constructor(node, ast) {
        this.node = node;
        this.ast = ast;
        const parameters = this.node.parameters;
        for (const parameter of parameters) {
            this.arguments[parameter.name.getText()] = new DaggerArgument(parameter, this.ast);
        }
    }
    getArgsOrder() {
        return Object.keys(this.arguments);
    }
    getReferences() {
        const references = [];
        for (const argument of Object.values(this.arguments)) {
            const ref = argument.getReference();
            if (ref) {
                references.push(ref);
            }
        }
        return references;
    }
    propagateReferences(references) {
        for (const argument of Object.values(this.arguments)) {
            argument.propagateReferences(references);
        }
    }
    toJSON() {
        return {
            arguments: this.arguments,
        };
    }
}
