// @dagger.io/dagger/dist/src/module/introspector/dagger_module/enumClass.js
import ts from "typescript";
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST } from "../typescript_module/index.js";
import { Locatable } from "./locatable.js";
export class DaggerEnumClassValue extends Locatable {
    node;
    ast;
    name;
    value;
    description;
    deprecated;
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        const initializer = this.node.initializer;
        if (!initializer) {
            throw new Error("Dagger enum value has no value set");
        }
        this.value = this.ast.resolveParameterDefaultValue(initializer);
    }
    toJSON() {
        return {
            name: this.name,
            value: this.value,
            description: this.description,
            deprecated: this.deprecated,
        };
    }
}
export class DaggerEnumClass extends Locatable {
    node;
    ast;
    name;
    description;
    values = {};
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of enum at ${AST.getNodePosition(node)}.`);
        }
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        this.description = this.ast.getDocFromSymbol(this.symbol);
        const properties = this.node.members;
        for (const property of properties) {
            if (ts.isPropertyDeclaration(property)) {
                const value = new DaggerEnumClassValue(property, this.ast);
                this.values[value.name] = value;
            }
        }
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            values: this.values,
        };
    }
}
