// @dagger.io/dagger/dist/src/module/introspector/dagger_module/enum.js
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST } from "../typescript_module/index.js";
import { Locatable } from "./locatable.js";
export class DaggerEnumValue extends Locatable {
    node;
    ast;
    name;
    value;
    description;
    deprecated;
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        this.name = this.node.name.getText();
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        const initializer = this.node.initializer;
        if (!initializer) {
            throw new IntrospectionError(`enum ${this.name} at ${AST.getNodePosition(this.node)} has no value set to its member.`);
        }
        this.value = this.ast.resolveParameterDefaultValue(initializer);
    }
    toJSON() {
        return {
            name: this.name,
            value: this.value,
            description: this.description,
            deprecated: this.deprecated,
        };
    }
}
export class DaggerEnum extends Locatable {
    node;
    ast;
    name;
    description;
    values = {};
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        this.description = this.ast.getDocFromSymbol(this.symbol);
        const members = this.node.members;
        for (const member of members) {
            const value = new DaggerEnumValue(member, this.ast);
            this.values[value.name] = value;
        }
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            values: this.values,
        };
    }
}
