// @dagger.io/dagger/dist/src/module/introspector/dagger_module/function.js
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST, isTypeDefResolved, resolveTypeDef, } from "../typescript_module/index.js";
import { DaggerArgument } from "./argument.js";
import { CHECK_DECORATOR, FUNCTION_DECORATOR, GENERATOR_DECORATOR, } from "./decorator.js";
import { Locatable } from "./locatable.js";
export class DaggerFunction extends Locatable {
    node;
    ast;
    name;
    description;
    deprecated;
    _returnTypeRef;
    returnType;
    arguments = {};
    alias;
    cache;
    isCheck = false;
    isGenerator = false;
    signature;
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        this.symbol = this.ast.getSymbolOrThrow(node.name);
        this.signature = this.ast.getSignatureFromFunctionOrThrow(node);
        this.name = this.node.name.getText();
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        const functionArguments = this.ast.getDecoratorArgument(this.node, FUNCTION_DECORATOR, "object");
        if (functionArguments) {
            if (typeof functionArguments === "string") {
                // previously only a single arg was accepted for alias, so if there's just
                // a string we intrepret it that way for backward compatibility
                this.alias = functionArguments;
            }
            else {
                this.alias = functionArguments.alias;
                this.cache = functionArguments.cache;
            }
        }
        // Parse @check decorator
        if (this.ast.isNodeDecoratedWith(this.node, CHECK_DECORATOR)) {
            this.isCheck = true;
        }
        // Parse @generate decorator
        if (this.ast.isNodeDecoratedWith(this.node, GENERATOR_DECORATOR)) {
            this.isGenerator = true;
        }
        for (const parameter of this.node.parameters) {
            this.arguments[parameter.name.getText()] = new DaggerArgument(parameter, this.ast);
        }
        this.returnType = this.getReturnType();
    }
    getReturnType() {
        const type = this.signature.getReturnType();
        const typedef = this.ast.tsTypeToTypeDef(this.node, type);
        if (typedef === undefined || !isTypeDefResolved(typedef)) {
            this._returnTypeRef = this.ast.typeToStringType(type);
        }
        return typedef;
    }
    getArgsOrder() {
        return Object.keys(this.arguments);
    }
    getReferences() {
        const references = [];
        if (this._returnTypeRef &&
            (this.returnType === undefined || !isTypeDefResolved(this.returnType))) {
            references.push(this._returnTypeRef);
        }
        for (const argument of Object.values(this.arguments)) {
            const reference = argument.getReference();
            if (reference) {
                references.push(reference);
            }
        }
        return references;
    }
    propagateReferences(references) {
        for (const argument of Object.values(this.arguments)) {
            argument.propagateReferences(references);
        }
        if (!this._returnTypeRef) {
            return;
        }
        if (this.returnType && isTypeDefResolved(this.returnType)) {
            return;
        }
        const typeDef = references[this._returnTypeRef];
        if (!typeDef) {
            throw new IntrospectionError(`could not find type reference for ${this._returnTypeRef} at ${AST.getNodePosition(this.node)}.`);
        }
        this.returnType = resolveTypeDef(this.returnType, typeDef);
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            deprecated: this.deprecated,
            alias: this.alias,
            arguments: this.arguments,
            returnType: this.returnType,
        };
    }
}
