// @dagger.io/dagger/dist/src/module/introspector/dagger_module/interfaceFunction.js
import ts from "typescript";
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST, isTypeDefResolved, resolveTypeDef, } from "../typescript_module/index.js";
import { DaggerArgument } from "./argument.js";
import { Locatable } from "./locatable.js";
export class DaggerInterfaceFunction extends Locatable {
    node;
    ast;
    name;
    description;
    deprecated;
    _returnTypeRef;
    returnType;
    arguments = {};
    symbol;
    signature;
    // Just placeholders to be compatible with `Method` during registration
    alias;
    cache;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of interface function at ${AST.getNodePosition(node)}`);
        }
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        // If it's a method signature, we can directly use it to get the signature props
        // If it's a property kind signature, we need to get the signature from the type
        // Examle:
        // interface Foo {
        //   bar(): void       // <- this is a method signature
        //   baz: () => string // <- this is a property kind signature
        // }
        const nodeType = this.node.type && ts.isFunctionTypeNode(this.node.type)
            ? this.node.type
            : this.node;
        const signature = this.ast.getSignatureFromFunctionOrThrow(nodeType);
        for (const parameter of nodeType.parameters) {
            this.arguments[parameter.name.getText()] = new DaggerArgument(parameter, this.ast);
        }
        const signatureReturnType = signature.getReturnType();
        const typedef = this.ast.tsTypeToTypeDef(this.node, signatureReturnType);
        if (typedef === undefined || !isTypeDefResolved(typedef)) {
            this._returnTypeRef = this.ast.typeToStringType(signatureReturnType);
        }
        this.returnType = typedef;
    }
    getReferences() {
        const references = [];
        if (this._returnTypeRef &&
            (this.returnType === undefined || !isTypeDefResolved(this.returnType))) {
            references.push(this._returnTypeRef);
        }
        for (const argument of Object.values(this.arguments)) {
            const reference = argument.getReference();
            if (reference) {
                references.push(reference);
            }
        }
        return references;
    }
    propagateReferences(references) {
        for (const argument of Object.values(this.arguments)) {
            argument.propagateReferences(references);
        }
        if (!this._returnTypeRef) {
            return;
        }
        if (this.returnType && isTypeDefResolved(this.returnType)) {
            return;
        }
        const typeDef = references[this._returnTypeRef];
        if (!typeDef) {
            throw new IntrospectionError(`could not find type reference for ${this._returnTypeRef} at ${AST.getNodePosition(this.node)}.`);
        }
        this.returnType = resolveTypeDef(this.returnType, typeDef);
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            deprecated: this.deprecated,
            arguments: this.arguments,
            returnType: this.returnType,
        };
    }
}
