// @dagger.io/dagger/dist/src/module/introspector/dagger_module/interface.js
import ts from "typescript";
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST } from "../typescript_module/index.js";
import { DaggerInterfaceFunction, } from "./interfaceFunction.js";
import { Locatable } from "./locatable.js";
export class DaggerInterface extends Locatable {
    node;
    ast;
    name;
    description;
    functions = {};
    symbol;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of interface at ${AST.getNodePosition(node)}`);
        }
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        this.description = this.ast.getDocFromSymbol(this.symbol);
        for (const member of this.node.members) {
            if (!ts.isPropertySignature(member) && !ts.isMethodSignature(member)) {
                continue;
            }
            // Check if it's a function
            if ((member.type && ts.isFunctionTypeNode(member.type)) ||
                ts.isMethodSignature(member)) {
                const daggerInterfaceFunction = new DaggerInterfaceFunction(member, this.ast);
                this.functions[daggerInterfaceFunction.name] = daggerInterfaceFunction;
                continue;
            }
            // TODO(TomChv): Add support for fields
        }
    }
    getReferences() {
        const references = [];
        for (const fn of Object.values(this.functions)) {
            references.push(...fn.getReferences());
        }
        return references.filter((v, i, arr) => arr.indexOf(v) === i);
    }
    propagateReferences(references) {
        for (const fn of Object.values(this.functions)) {
            fn.propagateReferences(references);
        }
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            functions: this.functions,
        };
    }
}
