// @dagger.io/dagger/dist/src/module/introspector/dagger_module/object.js
import ts from "typescript";
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST } from "../typescript_module/index.js";
import { DaggerConstructor } from "./constructor.js";
import { FUNCTION_DECORATOR, OBJECT_DECORATOR } from "./decorator.js";
import { DaggerFunction } from "./function.js";
import { Locatable } from "./locatable.js";
import { DaggerProperty } from "./property.js";
/**
 * Represents an object defined using the `class` keyword.
 *
 * The class may contains methods and fields, that may or may not be exposed to the Dagger API.
 *
 * @example
 * ```ts
 * type MyObject = {
 *   name: string
 *   age: number
 * }
 * ```
 */
export class DaggerObject extends Locatable {
    node;
    ast;
    name;
    description;
    deprecated;
    _constructor = undefined;
    methods = {};
    properties = {};
    symbol;
    kind() {
        return "class";
    }
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of class at ${AST.getNodePosition(node)}.`);
        }
        this.name = this.node.name.getText();
        if (!this.ast.isNodeDecoratedWith(node, OBJECT_DECORATOR)) {
            throw new IntrospectionError(`class ${this.name} at ${AST.getNodePosition(node)} is used by the module but not exposed with a dagger decorator.`);
        }
        const modifiers = ts.getCombinedModifierFlags(this.node);
        if (!(modifiers & ts.ModifierFlags.Export)) {
            console.warn(`missing export in class ${this.name} at ${AST.getNodePosition(node)} but it's used by the module.`);
        }
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        for (const member of this.node.members) {
            if (ts.isPropertyDeclaration(member)) {
                const property = new DaggerProperty(member, this.ast);
                this.properties[property.alias ?? property.name] = property;
                continue;
            }
            if (ts.isConstructorDeclaration(member)) {
                this._constructor = new DaggerConstructor(member, this.ast);
                continue;
            }
            if (ts.isMethodDeclaration(member) &&
                this.ast.isNodeDecoratedWith(member, FUNCTION_DECORATOR)) {
                const daggerFunction = new DaggerFunction(member, this.ast);
                this.methods[daggerFunction.alias ?? daggerFunction.name] =
                    daggerFunction;
                continue;
            }
        }
    }
    getLocation() {
        return AST.getNodeLocation(this.node);
    }
    getReferences() {
        const references = [];
        if (this._constructor) {
            references.push(...this._constructor.getReferences());
        }
        for (const property of Object.values(this.properties)) {
            const ref = property.getReference();
            if (ref) {
                references.push(ref);
            }
        }
        for (const fn of Object.values(this.methods)) {
            references.push(...fn.getReferences());
        }
        return references.filter((v, i, arr) => arr.indexOf(v) === i);
    }
    propagateReferences(references) {
        if (this._constructor) {
            this._constructor.propagateReferences(references);
        }
        for (const property of Object.values(this.properties)) {
            property.propagateReferences(references);
        }
        for (const fn of Object.values(this.methods)) {
            fn.propagateReferences(references);
        }
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            deprecated: this.deprecated,
            constructor: this._constructor,
            methods: this.methods,
            properties: this.properties,
        };
    }
}
