// @dagger.io/dagger/dist/src/module/introspector/dagger_module/property.js
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST, isTypeDefResolved, resolveTypeDef, } from "../typescript_module/index.js";
import { FIELD_DECORATOR, FUNCTION_DECORATOR } from "./decorator.js";
import { Locatable } from "./locatable.js";
export class DaggerProperty extends Locatable {
    node;
    ast;
    name;
    description;
    deprecated;
    alias;
    isExposed;
    symbol;
    _typeRef;
    type;
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of class at ${AST.getNodePosition(node)}.`);
        }
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        this.name = this.node.name.getText();
        this.isExposed =
            this.ast.isNodeDecoratedWith(this.node, FUNCTION_DECORATOR) ||
                this.ast.isNodeDecoratedWith(this.node, FIELD_DECORATOR);
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        this.alias = this.getAlias();
        this.type = this.getType();
    }
    getAlias() {
        let alias = this.ast.getDecoratorArgument(this.node, FUNCTION_DECORATOR, "string");
        if (alias) {
            return JSON.parse(alias.replace(/'/g, '"'));
        }
        alias = this.ast.getDecoratorArgument(this.node, FIELD_DECORATOR, "string");
        if (alias) {
            return JSON.parse(alias.replace(/'/g, '"'));
        }
    }
    getType() {
        const type = this.ast.checker.getTypeAtLocation(this.node);
        const typedef = this.ast.tsTypeToTypeDef(this.node, type);
        if (typedef === undefined || !isTypeDefResolved(typedef)) {
            this._typeRef = this.ast.typeToStringType(type);
        }
        return typedef;
    }
    getReference() {
        if (this._typeRef &&
            (this.type === undefined || !isTypeDefResolved(this.type))) {
            return this._typeRef;
        }
        return undefined;
    }
    propagateReferences(references) {
        if (!this._typeRef) {
            return;
        }
        if (this.type && isTypeDefResolved(this.type)) {
            return;
        }
        const typeDef = references[this._typeRef];
        if (!typeDef) {
            throw new IntrospectionError(`could not find type reference for ${this._typeRef} at ${AST.getNodePosition(this.node)}.`);
        }
        this.type = resolveTypeDef(this.type, typeDef);
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            deprecated: this.deprecated,
            alias: this.alias,
            type: this.type,
            isExposed: this.isExposed,
        };
    }
}
