// @dagger.io/dagger/dist/src/module/introspector/dagger_module/typeObject.js
import ts from "typescript";
import { IntrospectionError } from "../../../common/errors/index.js";
import { AST } from "../typescript_module/index.js";
import { Locatable } from "./locatable.js";
import { DaggerObjectTypeProperty, } from "./typeObjectProperty.js";
/**
 * Represents an object defined using the `type` keyword.
 *
 * Type object can only contains fields, no methods are allowed.
 * All fields are public and exposed to the Dagger API.
 *
 * @example
 * ```ts
 * @object()
 * export class MyObject {
 *  @func()
 *  public name: string
 *
 *  @func()
 *  async getName(): Promise<string> {
 *    return this.name
 *  }
 * }
 * ```
 */
export class DaggerTypeObject extends Locatable {
    node;
    ast;
    name;
    description;
    deprecated;
    _constructor = undefined;
    methods = {};
    properties = {};
    symbol;
    kind() {
        return "object";
    }
    constructor(node, ast) {
        super(node);
        this.node = node;
        this.ast = ast;
        if (!this.node.name) {
            throw new IntrospectionError(`could not resolve name of enum at ${AST.getNodePosition(node)}.`);
        }
        this.name = this.node.name.getText();
        this.symbol = this.ast.getSymbolOrThrow(this.node.name);
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        const type = this.ast.getTypeFromTypeAlias(this.node);
        if (type.flags & ts.TypeFlags.Object) {
            const objectType = type;
            const properties = objectType.getProperties();
            for (const property of properties) {
                const daggerProperty = new DaggerObjectTypeProperty(this.node, property, this.ast);
                this.properties[daggerProperty.name] = daggerProperty;
            }
        }
    }
    getLocation() {
        return AST.getNodeLocation(this.node);
    }
    getReferences() {
        const references = [];
        for (const property of Object.values(this.properties)) {
            const ref = property.getReference();
            if (ref) {
                references.push(ref);
            }
        }
        return references.filter((v, i, arr) => arr.indexOf(v) === i);
    }
    propagateReferences(references) {
        for (const property of Object.values(this.properties)) {
            property.propagateReferences(references);
        }
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            properties: this.properties,
            deprecated: this.deprecated,
        };
    }
}
