// @dagger.io/dagger/dist/src/module/introspector/dagger_module/typeObjectProperty.js
import { IntrospectionError } from "../../../common/errors/index.js";
import { isTypeDefResolved, resolveTypeDef, } from "../typescript_module/index.js";
import { Locatable } from "./locatable.js";
export class DaggerObjectTypeProperty extends Locatable {
    node;
    symbol;
    ast;
    name;
    description;
    deprecated;
    alias = undefined;
    isExposed = true;
    _typeRef;
    type;
    constructor(node, symbol, ast) {
        super(node);
        this.node = node;
        this.symbol = symbol;
        this.ast = ast;
        this.name = symbol.name;
        const { description, deprecated } = this.ast.getSymbolDoc(this.symbol);
        this.description = description;
        this.deprecated = deprecated;
        const type = this.ast.checker.getTypeOfSymbolAtLocation(this.symbol, this.node);
        this.type = this.ast.tsTypeToTypeDef(this.node, type);
        if (this.type === undefined || !isTypeDefResolved(this.type)) {
            this._typeRef = this.ast.typeToStringType(type);
        }
    }
    getReference() {
        if (this._typeRef &&
            (this.type === undefined || !isTypeDefResolved(this.type))) {
            return this._typeRef;
        }
        return undefined;
    }
    propagateReferences(references) {
        if (!this._typeRef) {
            return;
        }
        if (this.type && isTypeDefResolved(this.type)) {
            return;
        }
        const typeDef = references[this._typeRef];
        if (!typeDef) {
            throw new IntrospectionError(`could not find type reference for ${this._typeRef}.`);
        }
        this.type = resolveTypeDef(this.type, typeDef);
    }
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            deprecated: this.deprecated,
            alias: this.alias,
            type: this.type,
            isExposed: this.isExposed,
        };
    }
}
